File: naturereserve.py
from sys import stdin,stdout
from heapq import heappush,heappop,heapify

for _ in range(int(stdin.readline())):                   
    N,M,L,S = map(int,stdin.readline().split())
    frontier = list(map(lambda x: int(x) - 1,stdin.readline().split()))
    #run Prim's algorithm from the S stations and ensure n - S edges are chosen in O(M log N)
    #answer =  cost of N - S edges + (N - S) * L
    adjList,taken = [[] for _ in range(N)],[False for _ in range(N)]
    for i in range(M):
        u,v,w = map(int,stdin.readline().split())
        u -= 1; v -= 1
        adjList[u].append((v,w))
        adjList[v].append((u,w))
    #multi source Prim's algorithm
    cost,PQ,E = 0,[],0
    for u in frontier:
        for v,w in adjList[u]:
            PQ.append((w,v))
            taken[u] = True
    heapify(PQ)
    while PQ and E < N - S:
        w,u = heappop(PQ)
        if not taken[u]:
            cost += w
            E += 1
            taken[u] = True
            for v,w2 in adjList[u]:
                if not taken[v]:
                    heappush(PQ,(w2,v))
    stdout.write(f"{cost + (N - S) * L}\n")

File: naturereserve.cpp
#include <bits/stdc++.h>
#include <queue>

using namespace std;

typedef pair<int,int> ii;
typedef vector<int> vi;
typedef vector<ii> vii;
typedef vector<vii> AL;
typedef long long ll;

int main() {
    int D; scanf("%d",&D);
    while (D--) {
        ll N,M,L,S; scanf("%lld %lld %lld %lld",&N,&M,&L,&S);
        vi frontier; 
        for (int i = 0; i < S; i++) {
            int s; scanf("%d",&s);
            frontier.push_back(--s);
        }
        AL adjList(N, vii {}); vector<bool> taken(N, false);
        for (int i = 0; i < M; i++) {
            int u,v,w; scanf("%d %d %d",&u,&v,&w);
            u--; v--;
            adjList[u].push_back(make_pair(v,w));
            adjList[v].push_back(make_pair(u,w));
        }
        //run Prim's algorithm from the S stations and ensure n - S edges are chosen in O(M log N)
        //answer =  cost of N - S edges + (N - S) * L
        ll cost = 0; int E = 0;
        priority_queue<ii> PQ;
        //since C++ priority queue is max heap, negation is required to convert to min heap
        for (int u: frontier) {
            for (auto &[v,w]: adjList[u]) {
                PQ.emplace(-w,-v); // sort by increasing weight and increasing ID
                taken[u] = true;
            }
        }
        while (!PQ.empty() && E < N - S) {
            auto [w,u] = PQ.top(); PQ.pop();
            w = -w; u = -u;
            if (!taken[u]) {
                cost += w; E += 1;
                taken[u] = true;
                for (auto &[v,w2]: adjList[u]) {
                    if (!taken[v]) PQ.emplace(-w2,-v);
                }
            }
        }
        printf("%lld\n",cost + (N - S) * L);
    }
    return 0;
}